working/sealed-types/feature-specification.md

```diff
@@ -0,0 +1,196 @@
+# Sealed types
+
+Author: Bob Nystrom
+
+Status: In-progress
+
+Version 1.0
+
+This proposal specifies *sealed types*, which is core capability needed for
+[exhaustiveness checking][] of subtypes in [pattern matching][]. This proposal
+is a subset of the [type modifiers][] proposal. (We may wish to do all or parts
+of the rest of that proposal, but they aren't needed for pattern matching, so
+this proposal separates them out.) For motivation, see the previously linked
+documents.
+
+[exhaustiveness checking]: https://github.com/dart-lang/language/blob/master/working/0546-patterns/exhaustiveness.md
+
+[pattern matching]: https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md
+
+[type modifiers]: https://github.com/dart-lang/language/blob/master/working/type-modifiers/feature-specification.md
+
+## Introduction
+
+Marking a type `sealed` applies two restrictions:
+
+*   If it's a class, the type itself can't be directly constructed. The class is
+    implicitly `abstract`.
```

Comment:
What is the reasoning for this restriction?

I'm thinking that we *can* allow the superclass to be instantiated. Then we need to include that class itself in the set of types to check to achieve exhaustiveness. You can still write `abstract` to avoid it.
Disallowing it should have a reason. I'm sure there are plenty of good reason, I'm just curious which ones we've used to reach the conclusion :)

I can come up with:
* It's what you'd usually want. Avoids having to write `abstract` on the superclass.
* If not, people may forget to write `abstract` on the superclass, then their switches are not exhaustive.
* If we use `abstract` on superclass to decide whether it needs to be part of exhaustiveness checking, we are adding implicit meaning on top of just being `abstract`. That really should be explicit opt-in.
* If the superclass itself is part of exhaustiveness checking, ordering of switch cases matter. Breaks what should be nicely symmetric.
* It makes exhaustiveness checking easier to just disallow it.

Reply:
> What is the reasoning for this restriction?

It's necessary to make sealed types *useful*. If you can have direct instances of the supertype, then in pattern matching, you have to explicitly check the supertype in addition to its subtypes. But if you have to check the supertype anyway... then you derive absolutely no benefit from marking it `sealed` in the first place. The whole point of sealing is to know that if you've handled all the subtypes you are done. Direct instances of the supertype break that.